File: fsharp/interest-is-interesting/InterestIsInteresting.fs
module InterestIsInteresting

let interestRate (balance: decimal): single =
    match balance with
        | b when b < 0m -> 3.213f
        | b when b < 1000m -> 0.5f
        | b when b < 5000m -> 1.621f
        | _ -> 2.475f 

let interest (balance: decimal): decimal =
   balance * decimal(interestRate balance) / 100m

let annualBalanceUpdate(balance: decimal): decimal =
   balance + interest(balance)
   
let amountToDonate(balance: decimal) (taxFreePercentage: float): int =
    match balance with
        | b when b < 0m -> 0
        | _ -> float(balance) * taxFreePercentage * 2.0 / 100.0 |> int 

File: solutions/csharp/interest-is-interesting/2/InterestIsInteresting.cs
using System;

static class SavingsAccount
{
    public static float InterestRate(decimal balance)
    {
        return balance switch {
            < 0m => 3.213f,
            < 1000m => 0.5f,
            < 5000m => 1.621f,
            _ => 2.475f
        };
    }

    public static decimal Interest(decimal balance)
    {
        return (balance * Convert.ToDecimal(InterestRate(balance))) / 100m; 
    }

    public static decimal AnnualBalanceUpdate(decimal balance)
    {
        return balance + Interest(balance);
    }

    public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
    {
        int years = 0;
        decimal updatedBalance = balance;

        while (updatedBalance < targetBalance) {
            years++;
            updatedBalance = AnnualBalanceUpdate(updatedBalance);
        }
        return years;
    }
}
